test/stdlib/Diffing.swift

```diff
@@ -0,0 +1,667 @@
+
+  // Differs by type in the other direction:
+  expectNotEqual(
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 0, associatedWith: 0)
+  )
+
+  // Insert differs by offset
+  expectNotEqual(
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.insert(offset: 1, element: 0, associatedWith: 0)
+  )
+
+  // Insert differs by element
+  expectNotEqual(
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 1, associatedWith: 0)
+  )
+
+  // Insert differs by association
+  expectNotEqual(
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.insert(offset: 0, element: 0, associatedWith: 1)
+  )
+
+  // Remove differs by offset
+  expectNotEqual(
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.remove(offset: 1, element: 0, associatedWith: 0)
+  )
+
+  // Remove differs by element
+  expectNotEqual(
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 1, associatedWith: 0)
+  )
+
+  // Remove differs by association
+  expectNotEqual(
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 0, associatedWith: 0),
+    CollectionDifference<Int>.Change.remove(offset: 0, element: 0, associatedWith: 1)
+  )
+}
+
+suite.test("Compile-time test of hashable conformance") {
+  let _ = Set<CollectionDifference<String>>();
+}
+
+suite.test("Move inference") {
+  let n = CollectionDifference<String>.init([
+    .insert(offset: 3, element: "Sike", associatedWith: nil),
+    .insert(offset: 4, element: "Sike", associatedWith: nil),
+    .insert(offset: 2, element: "Hello", associatedWith: nil),
+    .remove(offset: 6, element: "Hello", associatedWith: nil),
+    .remove(offset: 8, element: "Goodbye", associatedWith: nil),
+    .remove(offset: 9, element: "Sike", associatedWith: nil),
+  ])
+  let w = CollectionDifference<String>.init([
+    .insert(offset: 3, element: "Sike", associatedWith: nil),
+    .insert(offset: 4, element: "Sike", associatedWith: nil),
+    .insert(offset: 2, element: "Hello", associatedWith: 6),
+    .remove(offset: 6, element: "Hello", associatedWith: 2),
+    .remove(offset: 8, element: "Goodbye", associatedWith: nil),
+    .remove(offset: 9, element: "Sike", associatedWith: nil),
+  ])
+  expectEqual(w, n?.inferringMoves())
+}
+
+suite.test("Three way diff demo code") {
+  let base = "Is\nit\ntime\nalready?"
+  let theirs = "Hi\nthere\nis\nit\ntime\nalready?"
+  let mine = "Is\nit\nreview\ntime\nalready?"
+
+  // Split the contents of the sources into lines
+  let baseLines = base.components(separatedBy: "\n")
```

Comment:
Isn't it an NSString API? How is it even available here without `import Foundation`, I wonder.

Reply:
Maybe through StdlibUnittest

Reply:
I've only run the tests on Darwin, it seems likely that Foundation is getting imported somewhere. I'll make the test more literal.